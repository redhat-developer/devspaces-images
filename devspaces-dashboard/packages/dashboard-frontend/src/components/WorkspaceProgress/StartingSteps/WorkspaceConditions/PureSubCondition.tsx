/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { ProgressStepTitle } from '@/components/WorkspaceProgress/StepTitle';

export type Props = {
  distance: -1 | 0 | 1 | undefined;
  title: string | undefined;
};

export class PureSubCondition extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    const { distance, title } = this.props;

    if (!title) {
      return <></>;
    }

    return (
      <ol className="pf-c-wizard__nav-list">
        <li className="pf-c-wizard__nav-item">
          <div className="pf-c-wizard__nav-link">
            <ProgressStepTitle distance={distance}>{title}</ProgressStepTitle>
          </div>
        </li>
      </ol>
    );
  }
}
