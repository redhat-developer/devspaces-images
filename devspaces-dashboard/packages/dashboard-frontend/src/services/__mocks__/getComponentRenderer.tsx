/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';

export * from '@testing-library/react';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
