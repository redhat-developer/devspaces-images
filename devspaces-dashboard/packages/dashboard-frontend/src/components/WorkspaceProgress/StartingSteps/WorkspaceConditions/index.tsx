/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import common from '@eclipse-che/common';
import { AlertVariant } from '@patternfly/react-core';
import isEqual from 'lodash/isEqual';
import React from 'react';

import {
  ProgressStep,
  ProgressStepProps,
  ProgressStepState,
} from '@/components/WorkspaceProgress/ProgressStep';
import styles from '@/components/WorkspaceProgress/StartingSteps/WorkspaceConditions/index.module.css';
import { PureSubCondition } from '@/components/WorkspaceProgress/StartingSteps/WorkspaceConditions/PureSubCondition';
import { ProgressStepTitle } from '@/components/WorkspaceProgress/StepTitle';
import {
  ConditionType,
  isConditionError,
  isConditionReady,
} from '@/components/WorkspaceProgress/utils';
import { WorkspaceParams } from '@/Routes/routes';
import { AlertItem, LoaderTab } from '@/services/helpers/types';

export type Props = ProgressStepProps & {
  condition: ConditionType;
  matchParams: WorkspaceParams;
};
export type State = ProgressStepState & {
  isWarning: boolean;
  isReady: boolean;
  condition: ConditionType;
  subConditionTitle: string;
};

export default class StartingStepWorkspaceConditions extends ProgressStep<Props, State> {
  private timerId: number | undefined;

  constructor(props: Props) {
    super(props);

    this.state = this.buildState(props, undefined, undefined);
  }

  protected get name(): string {
    return this.state.name;
  }

  private buildState(props: Props, prevProps: Props | undefined, state: State | undefined): State {
    const condition = props.condition;
    const prevCondition = prevProps?.condition;

    let name: string;
    let subConditionTitle: string;
    if (state === undefined) {
      name = condition.message || condition.type;
      subConditionTitle = '';
    } else {
      name = condition.message || state.name;
      subConditionTitle = state.subConditionTitle;
    }

    return {
      isReady: isConditionReady(condition, prevCondition),
      isWarning: isConditionError(condition, prevCondition),
      name,
      condition,
      subConditionTitle,
    };
  }

  private checkForSubCondition(condition: ConditionType): void {
    if (this.timerId !== undefined || condition.status !== 'False') {
      return;
    }

    // Show sub-condition only for DeploymentReady
    if (condition.type === 'DeploymentReady') {
      // Show sub-condition after 20 seconds
      this.timerId = window.setTimeout(() => {
        this.setState({
          subConditionTitle:
            'Downloading IDE binaries... (it can take a few minutes depending on your internet connection)',
        });
      }, 20000);
    }
  }

  public componentDidMount() {
    const state = this.buildState(this.props, undefined, this.state);
    this.setState(state);

    this.checkForSubCondition(state.condition);
  }

  public async componentDidUpdate(prevProps: Props) {
    const state = this.buildState(this.props, prevProps, this.state);
    this.setState(state);

    if (state.subConditionTitle === '') {
      this.checkForSubCondition(state.condition);
    }
  }

  public shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
    if (!isEqual(this.props.condition, nextProps.condition)) {
      return true;
    }

    if (!isEqual(this.state.condition, nextState.condition)) {
      return true;
    }

    if (this.state.subConditionTitle !== nextState.subConditionTitle) {
      return true;
    }

    return false;
  }

  public componentWillUnmount() {
    this.toDispose.dispose();
  }
  protected async runStep(): Promise<boolean> {
    return false;
  }

  protected handleRestart(alertKey: string, tab: LoaderTab): void {
    this.props.onHideError(alertKey);

    this.clearStepError();
    this.props.onRestart(tab);
  }

  protected buildAlertItem(error: Error): AlertItem {
    const key = this.name;
    return {
      key,
      title: 'Failed to open the workspace',
      variant: AlertVariant.danger,
      children: common.helpers.errors.getMessage(error),
      actionCallbacks: [
        {
          title: 'Restart',
          callback: () => this.handleRestart(key, LoaderTab.Progress),
        },
      ],
    };
  }

  render() {
    const { isWarning: isWarning, isReady, subConditionTitle } = this.state;
    const hasChildren = this.props.hasChildren || !!subConditionTitle;

    const distance = isReady ? 1 : 0;
    const isError = false;

    return (
      <React.Fragment>
        <ProgressStepTitle
          className={styles.conditionTitle}
          distance={distance}
          hasChildren={hasChildren}
          isError={isError}
          isWarning={isWarning}
        >
          {this.name}
          <PureSubCondition distance={distance} title={subConditionTitle} />
        </ProgressStepTitle>
      </React.Fragment>
    );
  }
}
