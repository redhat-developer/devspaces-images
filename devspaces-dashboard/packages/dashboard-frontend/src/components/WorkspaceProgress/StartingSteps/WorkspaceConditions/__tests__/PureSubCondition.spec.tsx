/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { PureSubCondition } from '@/components/WorkspaceProgress/StartingSteps/WorkspaceConditions/PureSubCondition';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

jest.mock('@/components/WorkspaceProgress/StepTitle');

describe('Starting sub-steps, checking rendering', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot with no title', () => {
    expect(createSnapshot(undefined)).toMatchSnapshot();
  });

  test('snapshot with a title', () => {
    expect(createSnapshot('sub-step test', 1)).toMatchSnapshot();
  });

  it('should show the title and a proper icon(depends on the distance value)', () => {
    let distanceElement: HTMLElement | null;
    let stepTitle: HTMLElement | null;

    const { reRenderComponent } = renderComponent('sub-step test 1', -1);

    distanceElement = screen.queryByTestId('distance');
    expect(distanceElement).not.toBeNull();
    expect(distanceElement).toHaveTextContent('-1');

    stepTitle = screen.queryByTestId('step-title');
    expect(stepTitle).not.toBeNull();
    expect(stepTitle).toHaveTextContent('sub-step test 1');

    reRenderComponent('sub-step test 2', 0);

    distanceElement = screen.queryByTestId('distance');
    expect(distanceElement).not.toBeNull();
    expect(distanceElement).toHaveTextContent('0');

    stepTitle = screen.queryByTestId('step-title');
    expect(stepTitle).not.toBeNull();
    expect(stepTitle).toHaveTextContent('sub-step test 2');

    reRenderComponent('sub-step test 3', 1);

    distanceElement = screen.queryByTestId('distance');
    expect(distanceElement).not.toBeNull();
    expect(distanceElement).toHaveTextContent('1');

    stepTitle = screen.queryByTestId('step-title');
    expect(stepTitle).not.toBeNull();
    expect(stepTitle).toHaveTextContent('sub-step test 3');

    reRenderComponent(undefined, 1);

    distanceElement = screen.queryByTestId('distance');
    expect(distanceElement).toBeNull();

    stepTitle = screen.queryByTestId('step-title');
    expect(stepTitle).toBeNull();
  });
});

function getComponent(title: string | undefined, distance?: -1 | 0 | 1): React.ReactElement {
  return <PureSubCondition title={title} distance={distance} />;
}
