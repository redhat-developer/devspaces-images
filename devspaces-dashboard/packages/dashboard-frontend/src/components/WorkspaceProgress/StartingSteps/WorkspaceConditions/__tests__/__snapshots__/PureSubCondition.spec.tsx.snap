// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Starting sub-steps, checking rendering snapshot with a title 1`] = `
<ol
  className="pf-c-wizard__nav-list"
>
  <li
    className="pf-c-wizard__nav-item"
  >
    <div
      className="pf-c-wizard__nav-link"
    >
      <div>
        <div
          data-testid="distance"
        >
          1
        </div>
        <span
          className="done"
          data-testid="step-title"
        >
          sub-step test
        </span>
      </div>
    </div>
  </li>
</ol>
`;

exports[`Starting sub-steps, checking rendering snapshot with no title 1`] = `null`;
